Translation validator for dynamo guards.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97964
* #97963
* #103611
* __->__ #102563

This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

ysiraichi committed:
Translation validator for dynamo guards.

This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

[ghstack-poisoned]

lezcano commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1362,6 +1371,458 @@ def rbinary_magic_impl(self, other):
+    # At the time of the translation, all free variables present in the
+    # SymPy expression being translated must be already mapped to a Z3
+    # integer variable.
+    class SympyToZ3:
```
This class can be streamlined by using `torch.utils._sympy.interp.py`.

ysiraichi committed:
Fix + using sympy_interp. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

ysiraichi committed:
Temporarily turing translation validation on by default. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

ysiraichi commented:
I temporarily turned translation validation on by default. **Reason:** so that CI will run the benchmarks with translation validation. Will turn it off before merge, though.

ezyang commented on requirements.txt:
```diff
@@ -14,3 +14,4 @@ filelock
 networkx
 jinja2
 fsspec
+z3-solver
```
No, we shouldn't put this requirement here. If you need it in CI add to to our build.sh script.

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -2925,9 +2925,11 @@ def convert(idx, x):
-                            hint=x
+                            shape_env.create_symbol(x, source),
+                            hint=x,
+                            source=source
```
nice catch

ezyang commented on torch/_dynamo/config.py:
```diff
@@ -230,6 +230,9 @@
 numpy_ndarray_as_tensor = False
 
+# Uses z3 for validating the guard optimizations transformations.
+translation_validation = True
```
Should be False by default, methinks

lezcano replied:
cf. https://github.com/pytorch/pytorch/pull/102563#issuecomment-1574434446

ysiraichi committed:
Fix CI issues. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

ysiraichi commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1362,6 +1371,433 @@ def rbinary_magic_impl(self, other):
+    # about it correctly. e.g. division depends on the denominator being
+    # different than 0.
+    #
+    # This class helps propagate such an assertion to the caller.
```
For example, `div(a, b)` will only work correctly if we add `b != 0` to Z3 solver. So, we need to propagate these new assertions to the validator. It's basically a named tuple.

ezyang replied:
Could we guard on this whenever we do division so you don't need to track manually?

ysiraichi replied:
I think it would work for the input guards, when we are creating the FX nodes. But the only way I could see it working for output guards would be to call `self.shape_env.evaluate_expr` from the overridden `SymNode` operations (inside `_make_node_magic`). What do you think?

ezyang replied:
I think it is OK (good, even) to guard on `y != 0` when computing `x / y`.

ysiraichi committed:
Simplify FX graph creation + CI fixes. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

ysiraichi committed:
Fix lint errors. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

ysiraichi committed:
Add comment. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

ysiraichi commented:
**Update:**
- Stop using `Tracer`. Using `Graph` directly, instead
- Creating a dummy return node when `ShapeEnv` is instantiated (needed for deepcopy calls)
- Linting fixes

ysiraichi committed:
Improve python types. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

ysiraichi committed:
Fix lint + checking for guard suppression. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ysiraichi committed:
Add warning logs temporarily. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ysiraichi commented:
Even after rebasing, I still keep hitting this issue. Can't reproduce it on my own machine. Do you have any idea why this is happening? 

```python
______ TestImports.test_no_mutate_global_logging_on_import_path_functorch ______
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/test_testing.py", line 2219, in test_no_mutate_global_logging_on_import
    self.assertEqual(out.strip(), expected)
  File "/opt/conda/envs/py_3.11/lib/python3.11/site-packages/torch/testing/_internal/common_utils.py", line 3096, in assertEqual
    raise error_metas[0].to_error(
AssertionError: String comparison failed: '/opt/conda/envs/py_3.11/lib/python3.11/si[1022 chars]wxyz' != 'abcdefghijklmnopqrstuvwxyz'
```

ysiraichi commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2388,6 +2982,15 @@ def hint(s):
+                    if srcname in self.source_to_symbol:
+                        r = sympy.Eq(self.source_to_symbol[srcname], expr)
+                        self._add_output_guard(r)
+                    elif srcname not in concrete_sources:
+                        self.log.warning("source not found: %s (== %s)", srcname, expr)
```
Temporar warnings for identifying cases when: there is a source that reaches `produce_guards`, but does not go through `create_symintnode`.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -594,7 +595,7 @@ class SymNode:
     End users don't touch this.  Magic methods are NOT defined on this object.
     """
-    def __init__(self, expr, shape_env, pytype, hint: Optional[Union[int, float]], constant=None):
+    def __init__(self, expr, shape_env, pytype, hint: Optional[Union[int, float]], constant=None, fxnode=None):
```
nittiest of nits, but the spelling fxnode rankles me cuz it feels like it should be `fx_node` lol

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+    Z3Expr = Union[z3.ArithRef, z3.BoolRef]
+
+    def _boolref(expr: Any) -> z3.BoolRef:
+        return cast(z3.BoolRef, expr)
```
Not entirely sure what this does, esp since cast doesn't have any runtime effect

ezyang replied:
ok I guess this is just a convenient cast alias. Do you want to just assert isinstance(expr, z3.BoolRef) instead? Would be safer.

ysiraichi replied:
Killed with `Z3Node`

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+
+        @staticmethod
+        def floor(number: z3.ArithRef) -> Z3Node:
+            return Z3Node(Z3Ops.to_int(number).expr)
```
It is interesting you just drop the assertion here. In this case I suppose you know assertion is None because a division didn't happen, but it's kind of non-uniform (and if you ever modified the assertion to hold more interesting stuff, this could lead to bugs). I'm still more keen on separately guarding on not zero IMO.

ysiraichi replied:
What if, instead, `Z3Ops` held a reference to a `TranslationValidator` object? So that we would access it like: `Z3Ops(validator).<operator>(<args>)`?

ezyang replied:
sure

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+        @staticmethod
+        def floordiv(numerator: z3.ArithRef, denominator: z3.ArithRef) -> Z3Node:
+            div = Z3Ops.div(numerator, denominator)
+            return Z3Node(Z3Ops.to_int(div.arith).expr, div.assertion)
```
Does this handle negative rounding correctly?

ysiraichi replied:
I think so. According to [its spec](https://smtlib.cs.uiowa.edu/theories-Reals_Ints.shtml): _"`to_int` as the function that maps each real number `r` to its integer part, that is, to the largest integer `n` that satisfies `(<= (to_real n) r)`"_, i.e. rounds towards negative infinity.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+        @staticmethod
+        def ceil(number: z3.ArithRef) -> Z3Node:
+            ite = z3.If(Z3Ops.floor(number).arith < number, Z3Ops.floor(number + 1).arith, number)
+            return Z3Node(cast(z3.ArithRef, ite))
```
Interesting, is ArithRef really the accurate type annotation for nodes then?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+        def pow(base: z3.ArithRef, exp: z3.ArithRef) -> Z3Node:
+            return Z3Node(
+                expr=base ** exp,
+                assertion=_boolref(z3.Or(base != 0, exp > 0)),
```
Blegh, if you do the guard thing I suggested for assertion, we'd have to start generating or guards in PyTorch. I suppose it's all feasible but these will be difficult to simplify away (unless we statically can discharge them)

ysiraichi replied:
Hmm. Instead, we could just issue it to the validator, but outside this function (inside `make_node_magic` function).

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+    #   3. Lifts the return statement into 'Z3Node'
+    def _operator_to_z3(op: Callable) -> Callable:
+        # Operations that have booleans as their argument.
+        # This is neeeded because the argument of some FX nodes were
```
needed

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+                # types below. Finally, wrap it into a 'Z3Node'.
+                if isinstance(a, bool) or (as_bool and isinstance(a, int)):
+                    return Z3Node(z3.BoolVal(bool(a)))
+                if isinstance(a, (int, sympy.Integer)):
```
it's not clear to me when you could ever get a sympy.Integer here

ysiraichi replied:
Inside the function `evaluate_expr`, the variable `concrete_val` might be a `sympy.Integer`. So, we may end up creating an FX node like `operator.eq(expr, sympy.Integer(X))`.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+            sym_sqrt: lift(Z3Ops.sqrt),
+            # Not lifted because we only use this function as a
+            # marker for adding the expression as validator input.
+            torch._assert: torch._assert,
```
Huh, weird, wasn't expecting to see this here

ezyang replied:
Oh I see, its how you do the processing. okey

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+                # Ignore output nodes.
+                pass
+            else:
+                raise AssertionError(f"unsupported operation: {node.op}")
```
It's more idiomatic to use fx.Interpreter to abstract out this boilerplate!

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+
+            assert callable(node.target)
+            node_to_z3[node] = node.target(*args)
+            validator.add_assertion(node_to_z3[node].assertion)
```
Once you've added the assertion you don't need to keep them around in `node_to_z3` anymore, right?

ysiraichi replied:
I guess you are right. But if I only store the Z3 expression, I will also have to modify how I am lifting the operators. From `(Z3Node, ...) -> Z3Node` to `(Z3Expr, ...) -> Z3Node`.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+        def run_assertion(node: torch.fx.Node) -> None:
+            assertion = node.args[0]
+            assert isinstance(assertion, torch.fx.Node)
+            validator.add_input(node_to_z3[assertion].bool)
```
It's a bit weird for this to be called input. Maybe?

ysiraichi replied:
Agreed. I also find myself getting confused more often than not.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+                return self.Z3_REPLACEMENT_HANDLES[name]
+            if name in self.OPERATOR_HANDLES:
+                return getattr(operator, name)
+            print(name)
```
put this in the AttributeError

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+    #
+    # If the above equation is true, it means that there is a case where the
+    # generated guard (using the output expressions) is incorrectly true.
+    # Otherwise, it shows that the optimized output expressions is sound.
```
This can probably be written in a shorter form :)

ysiraichi replied:
I will try to summarize this comment.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+    # Otherwise, it shows that the optimized output expressions is sound.
+    class TranslationValidator:
+        def __init__(self) -> None:
+            super().__init__()
```
No superclass, this does nothing

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+
+            # Set of Z3 expressions representing the actual issued guards.
+            # i.e. those that will actually be used in the guard function.
+            self._outputs = set()
```
It might be better to stop using the input/output naming and do something like original/optimized

ezyang replied:
LLVM Alive2 uses "source" and "target" nomenclature.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+
+        # Retrieves the corresponding Z3 variable.
+        def z3var(self, symbol: sympy.Symbol) -> z3.ArithRef:
+            assert symbol in self.symbols
```
worth putting an extra message in assert, saying what the symbol was and what available symbols were

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+
+        # Checks whether all symbols were already added.
+        def _check_freesymbols(self, e: sympy.Basic) -> None:
+            assert all(isinstance(s, sympy.Symbol) and s in self.symbols for s in e.free_symbols)
```
Nit: better to do this in a loop so you can more easily give a better message when the assert fails. No big deal, easy refactor when it matters, just something to think about when making these asserts

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+            self._outputs.add(self.to_z3(e))
+
+        def add_assertion(self, e: Any) -> None:
+            if e is None:
```
odd, when does this happen?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+            model: Optional[z3.ModelRef] = None
+
+            # List of the expressions that failed due to the assignment.
+            failed_inputs: Optional[List] = None
```
nit: put in a parameter here for List

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+            failed_inputs: Optional[List] = None
+
+            # List of output expressions
+            outputs: Optional[Set] = None
```
ditto

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1929,9 +2390,97 @@ def __init__(
         self.frozen = False
         self.dim_constraints: Optional[DimConstraints] = None
 
+        self.graph_added_nodes: Dict[Tuple[Callable, Tuple[Any, ...]], torch.fx.Node] = {}
```
Not sure what this means, the type signature didn't help either

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1929,9 +2390,97 @@ def __init__(
         self.frozen = True
 
+    def _create_symbol_for_source(self, source: Source, positive: Optional[bool]) -> Optional[sympy.Symbol]:
+        if _translation_validator_enabled():
```
nit: avoid indentation by using early return instead

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1929,9 +2390,97 @@ def __init__(
+            srcname = source.name()
+            if source not in self.source_to_symbol:
+                self.source_to_symbol[srcname] = sympy.Symbol(srcname, positive=positive, integer=True)
+            return self.source_to_symbol[srcname]
```
So source_to_symbol is only populated during TV?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1929,9 +2390,97 @@ def __init__(
+        if _translation_validator_enabled():
+            srcname = source.name()
+            if source not in self.source_to_symbol:
+                self.source_to_symbol[srcname] = sympy.Symbol(srcname, positive=positive, integer=True)
```
Are you sure you want to allocate a fresh sympy Symbol here??

ysiraichi replied:
Well, that's the idea. This is so I make sure I have one unique symbol per source.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1929,9 +2390,97 @@ def __init__(
+            op: Callable,
+            args: Tuple,
+    ) -> Optional[torch.fx.Node]:
+        # Cache this tuple in order to avoid dupplicated nodes.
```
duplicated

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1929,9 +2390,97 @@ def __init__(
+        node_key = (op, args)
+
+        if _translation_validator_enabled() and node_key not in self.graph_added_nodes:
+            # Presence of None in the arguments implies that we should ignore this operation.
```
How come?

ysiraichi replied:
There are some reasons for that:
- Inductor creates `SymInt` nodes using `ShapeEnv`
- As far as I understand, we don't want to keep track of those
- We are not dealing with `None`, so we can use that as a marker of "uninteresting nodes"

ezyang replied:
Ah OK, I don't really understand the inductor fresh symints, but I agree this is a pain

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2037,49 +2587,86 @@ def create_symbolic_sizes_strides_storage_offset(
 
-        sym_sizes = [self.create_symintnode(i, hint=hint) for i, hint in zip(size, ex.size())]
+        sym_sizes = [
+            self.create_symintnode(sym, hint=hint, source=TensorPropertySource(source, TensorProperty.SIZE, i), positive=True)
```
Does positive include zero? Because size can defo be zero.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2037,49 +2587,86 @@ def create_symbolic_sizes_strides_storage_offset(
+            positive: Optional[bool] = None
+    ):
+        if _translation_validator_enabled() and source is not None:
+            # Create a new symbol for this source.
```
I can believe this is the best way to do it, but I'm a little fuzzy on the necessity of creating a new symbol and then asserting equality. Is this to prevent simplifications from happening to the input symbols?

ezyang replied:
If so, it seems important to have a clear nomenclature that distinguishes these two cases, and talk through how these are treated differently in ShapeEnv

ysiraichi replied:
As far as I understand, `sym` may not always be a `sympy.Symbol`. That's why we create a symbol for this source, and then assert equality.

ezyang replied:
specifically, it may just be an int, is that right?

ezyang commented on test/dynamo/test_misc.py:
```diff
@@ -5639,6 +5645,121 @@ def fn(x):
+
+        # Solutions for output is a subset of the solutions for the input.
+        validator.add_output(s0 > 20)
+        validator.add_output(s1 > s0 + 2)
```
Would be good to exercise cases where things got constant-ified

ysiraichi replied:
Not sure I understand what you mean by that. Is something like this what you are asking for?

```python
def f(t):
    assert t.shape[0] == 10
    return t

tensor = make_fx(f, tracing_mode="symbolic")(torch.randn(10))
self.assertExpectedInline(show_guards(tensor), """""")
```

ezyang replied:
yup, exactly that!

ezyang reviewed (approved):
Nice work!

ezyang commented:
As a reminder, I would love it if we could have TV permanently turned on in the test suite and the benchmark suite accuracy tests.

ysiraichi commented:
Right. I will look into that.

ysiraichi committed:
Update on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ysiraichi commented:
Here is a summary of the last changes:
- Renamed:
    - `fxnode` to `fx_node`
    - `TranslationValidator._inputs` to `TranslationValidator._source_exprs`
    - `TranslationValidator._outputs` to `TranslationValidator._target_exprs`
    - `ShapeEnv.graph_added_nodes` to `ShapeEnv.fx_node_cache`
- Using `z3.ExprRef` instead of `Z3Expr` (killed)
- Using `fx.Interpreter` instead of ad-hoc function for walking an FX graph
- `Z3Ops` holds a reference to a `TranslationValidator` for adding assertions
    - `Z3Node` was killed
- Logging `z3.unknown` results (i.e. when a solution is not found)
- Merged `TranslationValidator.add_<type>` into `add_var`
- Caching FX graph placeholders
- Casting `FloorDiv` results to real
    - Conforming to Python semantics

ysiraichi committed:
Revert translation validation by default. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ysiraichi committed:
Move Z3 CI dependency to first PR. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ysiraichi commented:
I would really appreciate that...
I tried logging into one of the CIs that were failing. Apparently, I am getting the following, instead of just `abcdefghijklmnopqrstuvwxyz `:

```python
/opt/conda/envs/py_3.11/lib/python3.11/site-packages/pkg_resources/__init__.py:121: DeprecationWarning: pkg_resources is deprecated as an API
  warnings.warn("pkg_resources is deprecated as an API", DeprecationWarning)
/opt/conda/envs/py_3.11/lib/python3.11/site-packages/pkg_resources/__init__.py:2870: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
  declare_namespace(pkg)
/opt/conda/envs/py_3.11/lib/python3.11/site-packages/pkg_resources/__init__.py:2870: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
  declare_namespace(pkg)
abcdefghijklmnopqrstuvwxyz
```

That said, I can't reproduce it on my machine.

ysiraichi commented:
Nvm. I think I got it. I am moving Z3 code to its own file (as suggested by @lezcano before).

ysiraichi committed:
Move validator into its own file. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ysiraichi committed:
Add test for specialization. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

[ghstack-poisoned]

ezyang commented:
@pytorchbot rebase

ysiraichi commented:
I will move the tests + CI changes to #103611 for consistency reasons.

ysiraichi committed:
Move tests to #103611. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi commented on torch/fx/experimental/validator.py:
```diff
@@ -0,0 +1,419 @@
+                    target_exprs=self._target_exprs,
+                )
+            else:
+                if r == z3.unknown:
+                    # Could not find a solution. It didn't fail, but it also
+                    # didn't succeed. Canceling the validation execution (keyboard
+                    # interrupt) also gets to this branch.
+                    log.warning("translation validation: could not validate")
+                else:
+                    # Target expressions are sound.
+                    assert r == z3.unsat
+                    log.debug("translation validation: success")
+                return self.Result(success=True)
```
On second thoughts, I don't think I should be doing this, but fail on `z3.unknown`.

ysiraichi replied:
@ezyang 
This makes lots of tests fail. What do you think?
Maybe we could land it as is, and try to tweak the validator later: e.g. find a better solver, use only reals...

ezyang replied:
Yeah, let's work on it more later

ezyang replied:
Maybe @nunoplopes has some ideas

nunoplopes replied:
You should ignore z3.unknown. That is either because the solver gave up or you reached a time/memory out, for example.
Tweaking the encoding and solver options is a whole different game. I would leave for a follow up patch. I can help with that ofc.
But I would try switching everything into bit-vectors; that's usually the most performing option.

ysiraichi replied:
Got it. So, for now, I will revert to the old behavior: issue a warning.

ysiraichi committed:
Handle unknowns correctly. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Fix success flag on z3.unknown. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi commented on torch/fx/experimental/validator.py:
```diff
@@ -0,0 +1,416 @@
+                    failed_source_expr=[inp for inp in self._source_exprs if not model.evaluate(inp)],
+                )
+            else:
+                if r == z3.unknown:
+                    # Could not find a solution. It didn't fail, but it also
+                    # didn't succeed. Canceling the validation execution (keyboard
+                    # interrupt) also gets to this branch.
+                    log.warning("translation validation: could not validate")
+                    return self.Result(success=False)
+                else:
+                    # Target expressions are sound.
+                    assert r == z3.unsat
+                    log.debug("translation validation: success")
+                    return self.Result(success=True)
```
Just realized I had gotten this backwards. I will wait for the CI just to see what happens if we fail on `z3.unknown`.

ysiraichi committed:
Skipping high-latency benchmarks. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi commented on torch/fx/experimental/validator.py:
```diff
@@ -0,0 +1,429 @@
+            return z3.If(
+                self.floor(number) < number,
+                self.floor(number + 1),
+                number
```
Z3 automatically casts `self.floor(number)` into real, if `number` is real.

ysiraichi committed:
Small fixes. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Kill temporary debugging code. on "Translation validator for dynamo guards."


This PR introduces a translation validator for dynamo guards. In summary, it verifies
whether the guards issued as Python code are sound, w.r.t the initial guards.

The main changes in this PR are:

- Create an FX graph for dynamic shapes
- Translate "the original" guards from the FX graph to Z3
- Check if the guards produced by `produce_guards` are sound w.r.t. the ones from the FX graph

gh-stack version of the PR #101146.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ysiraichi commented:
@pytorchbot merge
